lib/rucio/common/stomp_utils.py

```diff
@@ -15,145 +15,415 @@
 """
 Common utility functions for stomp connections
 """
+from __future__ import annotations
 
+import json
 import logging
+import random
 import socket
+from collections import namedtuple
+from copy import deepcopy
+from functools import partial
 from time import monotonic
 from typing import TYPE_CHECKING
 
-from stomp import Connection
+from stomp import Connection12
+from stomp.exception import ConnectFailedException, NotConnectedException
+from stomp.listener import HeartbeatListener
+
+from rucio.common.config import config_get, config_get_bool, config_get_float, config_get_int, config_get_list
+from rucio.common.logging import formatted_logger
+from rucio.core.monitor import MetricManager
 
 if TYPE_CHECKING:
-    from collections.abc import Callable, Sequence
-    from typing import Any
+    from collections.abc import Generator
+
+    from stomp.connect import Frame
+
+    from rucio.common.types import LoggerFunction
+
+
+METRICS = MetricManager(module=__name__)
+
+
+class Connection(Connection12):
+    """
+    Connection class.
+
+    Wraps Stomp Connection but knows the brokers without accessing
+    hidden variables from the Transport.
+    """
+    def __init__(self, host_and_ports: list[tuple[str, int]], **kwargs):
+        """
+        Initialise.
+
+        Args:
+            host_and_ports (list[tuple[str, int]]): brokers list
+
+        Kwargs:
+            Arguments to pass to the Constructor12 base class.
+        """
+        super().__init__(host_and_ports=host_and_ports, **kwargs)
+        self._brokers = host_and_ports
+
+    @property
+    def brokers(self) -> list[tuple[str, int]]:
+        """
+        List brokers.
+
+        Returns:
+            list[tuple[str, int]]: All assigned brokers in (host, port) format.
```

Comment:
Please avoid specifying the argument types in the docstring if they are already present as type annotations; there's no need to duplicate this information and it can cause mismatches between the type annotation and the type specified in the docstring if changes are made down the line